### PR TITLE
PROV: Relegate most of the FIPS provider code to libfips.a

### PR DESCRIPTION
**Pull request by levitte**
provider/fips/fipsprov.c contains a number of symbols that get used by
anything that's included in libfips.a, at least on Unix.
Unfortunately, there are platforms that do not support resolving
symbols to things that are already included in the end product (module
in this case) being built; they only support resolving symbols with
what comes next in the linking process.

The offending symbols in this case are FIPS_security_check_enabled,
c_thread_start and ossl_fips_intern_provider_init.

We resolve this by placing provider/fips/fipsprov.c in libfips.a along
with everything else there.  That takes care of the offending symbols.
What remains is to ensure that there is an entry point in an object
file used directly when linking the module, providers/fips/fips_entry.c

### CODE REVIEW

**Review by paulidale — APPROVED** [review]
One white space nit

File: providers/fips/build.info [hunk 1]
```diff
@@ -1,2 +1,6 @@
-SOURCE[../fips]=fipsprov.c self_test.c self_test_kats.c
-INCLUDE[../fips]=../implementations/include ../common/include ../..
+# We include the provider implementation into ../libfips.a, so that all
+# platforms can resolve symbols in other of that library's members.
+SOURCE[../libfips.a]=fipsprov.c self_test.c self_test_kats.c
+
+# It is necessary to have an explicit entry point
+SOURCE[../fips]=fips_entry.c
```
[the remark this answers is not in the document]
Comment (levitte) [reply providers/fips/build.info]: Fixed

**Review by paulidale — APPROVED** [review]

File: providers/fips/build.info [hunk 2]
```diff
@@ -1,2 +1,6 @@
-SOURCE[../fips]=fipsprov.c self_test.c self_test_kats.c
-INCLUDE[../fips]=../implementations/include ../common/include ../..
+# We include the provider implementation into ../libfips.a, so that all
+# platforms can resolve symbols in other of that library's members.
```
Comment (h-vetinari) [inline providers/fips/build.info:2]:
```suggestion
# platforms can resolve symbols in other members of that library.
```

Perhaps slightly more readable...

Comment (levitte) [reply providers/fips/build.info]: Done.  Thanks

**Comment by levitte** [comment]
There was a tweak, re-approval needed

File: providers/fips/fipsprov.c [hunk 3]
```diff
@@ -518,7 +518,22 @@ static const OSSL_DISPATCH intern_dispatch_table[] = {
...
+ * by the compiler if nothing else is said, but since this is part
+ * of libfips, and we build our libraries with mixed case symbol names,
+ * we must switch back to this default explicitly here.
+ */
+#ifdef __VMS
+# pragma names save
+# pragma names uppercase,truncated
+#endif
 OSSL_provider_init_fn OSSL_provider_init_int;
+#ifdef __VMS
```
Comment (paulidale) [inline providers/fips/fipsprov.c:534]: Should this be _after_ the function or does VMS remember changing the name after the restore?

Comment (levitte) [reply providers/fips/fipsprov.c:534]: Yes.  What this pragma does is basically to set a name treatment attribute on that *name*.  You'll notice that \<openssl/core.h\> does exactly the same for `OSSL_provider_init`.

**Review by paulidale — APPROVED** [review]

**Comment by levitte** [comment]
Merged

b938544969577e3b74da6f8c689c87c90ceced22 PROV: Relegate most of the FIPS provider code to libfips.a

**Pull request closed without merging**